### ISSUE TITLE
Breadth of application of customID

### ISSUE THREAD

**Issue by mc12345678** [potteryhouse/stock_by_attribute_1.5.4#16]
This was asked on the 1.5.3 version of this software for ZC 1.5.1, though is far more applicable to ZC 1.5.3 and this version of SBA.

Have found a way to pretty much remove SBA from the includes/classes/order.php file to use the notify system; however, there are two parts remaining to address. The one of "most" significance at this point is the customID. The reason this one is most difficult to address is that in all of the uses of the $this->products[$i]['model'] variable, there is only one location that the variable is substituted for the $customid and that point is products_ordered_html which is used to generate the html email.

But it is not used or substituted in low stock emails, product order history, or products_ordered (and therefore not in standard emails, but in html emails because of the inclusion in products_ordered_html.

Seems appropriate to include/incorporate the information into the regular/non-html emails; however, what about in the low stock emails and the tracked order history? Currently the method of substitution is if there is a customId then use the customid, otherwise use the model# (whether there or not)... So, the thought is if the customid is to replace all of the above, then it can be done with a relatively simple piece of code, then the last thing to work out in the order.php file above is generating the low-stock e-mail information that is applicable to attributes.

In the end if that can happen there are a total of three changes to be made to the includes/classes/order.php file to support SBA via notify/observer action. Would require adding one more variable declaration to capture the $order_products_attributes_id, a modification in the notifier just below it to pass it to the observer and one other notifier modification to pass the attribute stock to and from unless that variable can remain outside of order.php file and stay only inside the observer (supports the low stock email for a tracked attribute)...

So, question is, what is the objection to logging in the order history to use the customid if assigned or the model# if no customid?

**Comment by mc12345678** [comment]
Long and short of the above: the change logs the customid as the model# in the order history instead of the model# and some additional customid... The least intrusive to ZC would be to log both, but then I call to question the reporting of only one (customid if present otherwise model#) perhaps option should be available to report both rather than one or the other? 

Ideas? This is an open forum btw...

**Comment by jeking928** [comment]
I like the idea of logging both.  I can see situations where a shop owner would want the model# even when a customid exists.

**Comment by mc12345678** [comment]
Hey, there we go. So I'll change the way I set the logging of the order detail so that the customid is captchured in the new table and leave the order record for the model# intact. Codewise it will require the includes/classes/order.php to still be included in future versions of SBA, but at least all information will be maintained to be manipulated as the store owner sees fit.

Thanks for the input.

**Comment by jeking928** [comment]
I know this hasn't gotten a lot of attention.  I use it for two clients and neither needs the mixed attributes or customid. Thanks for keeping at it.

**Comment by mc12345678** [comment]
Regarding mixed attributes, most have come to know (think) that SBA can't handle them. I'm trying to take a turn on that. Currently it does work with Dynamic Dropdowns installed, though the template file should not be included as I've already merged it's action into the same template file for SBA.

Once I get a little more complete on how SBA will move forward and will post an update to that plugin to correct a few minor issues and maybe incorporate some of what is being done in SBA to support mixed attributes.

**Comment by jeking928** [comment]
I'll be sure to test it out once you do.

**Comment by mc12345678** [comment]
Hope you're ready to test... I haven't tried making a purchase again since I added the logging of the customid into the new orders table that tracks SBA specific order information, but I could use at least a second set of eyes.  There are many changes/additions to the most recent upload.  All files are in the branch: mc12345678_ZC154 which should be backwards compatible with ZC 1.5.3 as well. 

No pull request has been made yet.

**Comment by jeking928** [comment]
I'll give it a go.  Thanks for confirming which branch too.

**Comment by jeking928** [comment]
Installation when well in a vanilla 1.5.4 cart, once I fixed a typo.

stock_by_attributes_install.php
line 559, should be:  `customid` varchar(255) DEFAULT NULL,

Functionality is looking good. Out of Stock is grayed out, inventory debits correctly, I find no problems.

Oops, I take that back.  The receipt had the customID but the admin does not.
![image](https://cloud.githubusercontent.com/assets/8162029/6351786/f7484c0a-bc03-11e4-8cb2-1a9e0a0cfe49.png)

My concern is how the attributes are displayed. I used the Matrox G200 MMS and set stock and ID's for all six combinations of Model and Memory.  On the product page, the drop-down for for both Model and Memory contain six choices.  See screen shot:
![image](https://cloud.githubusercontent.com/assets/8162029/6351695/22291d24-bc03-11e4-82bc-1f89c3867114.png)
As a shopper, I find that confusing.  I don't know what the difference is between the three Value options.  The problem is not with SBA, but with know mixed-attributes are handled. Maybe I'm this is why Dynamic Dropdowns is needed?

**Comment by mc12345678** [comment]
Updated line 559 to be `customid` instead of 'customid' (quotes were incorrect)

Also found in trying to apply test customid's to confirm the order issue that the lower save button currently does not work on the admin page.  I did make a change earlier when "cleaning" up the page related to form controls and it appears that my modification disabled the lower save button.  Will have to look into what it takes to correct that and maintain the correctness of the html.

Still testing the customid aspect of the order to see if I can reproduce and what if anything needs to be modified to restore that feature.

As to the multiple choices... So yes, that is back to discussion somewhere in here about how to progress forward with multiple attributes...

Dynamic Dropdowns has already been developed, though its original state has a few things lacking.  1) if the database has a prefix then for some reason the code was put together to double the prefix (unnecessary and incorrect).
2) The typical/default installation of Dynamic Dropdowns requires all items with attributes to be tracked by SBA even if the attribute adds no stock value.  (That is addressed by the modified template file included in SBA so that if a product doesn't have an attribute tracked by SBA then the default ZC code takes over and there is no "issue".)
3) When working with multiple (mixed) attributes, it only produces results of options that are available and only in a sequential order (not a problem for it to be sequential, but there are those that would rather be able to let the customer choose from any of the available attributes and see what stock is available from those attributes that remain... Ie., if color of a shoe is the first choice, then would have to try each color to find a shoe that fits, whereas if size were chosen first (although listed second) then would be able to see all colors that are available).  Also, if a product is not currently available then there is no indication (other than listing in the product description) of what options are routinely available... So...

The sql was modified to provide the maximum results and still contain the quantity available.  If the sql statement were modified to not include quantity (and customid or other data specific to the selectable attribute) then single versions of each mixed attribute could be provided and the confusion would be non-existent, though availability would basically not be determinable until attempting to add to the cart.

Anyways, there are multiple methods to move forward (assuming the base product works), my first thought is to continue on using Dynamic Dropdowns and develop the necessary instruction to incorporate additional options into it rather than simply modifying SBA to use the actions and setup of Dynamic Dropdowns.  May be a bit of "patchwork", but I think that if an appropriate class is put together to support the continued use of Dynamic Dropdowns in parallel with SBA, then the two can continue to be maintained independently but installed together...  I believe some of this will be better put together through the use of admin controlled switches to address the various characteristics.  For example, if one does not care/need to show the customid and quantity of a specific variant, then SBA could provide a "basic" display of the product information without DDs, but as soon as either of those (or any additional variant specific information) is added for product display then we're back into handling/addressing the multiple rows returned.  Again it is all of these variations that have driven SBA to end up being a mess... Hopefully though they can all be put together into a single functional product.

**Comment by mc12345678** [comment]
I've still created an issue with the admin data entry that needs to be sorted.  Able to type in text in some cases/blocks, but not able to save it... :/

**Comment by jeking928** [comment]
I see a flaw in requiring DD, as a separate plugin, in order for SBA mixed attributes to work properly. I think DD would address most of the issues, but will be less than idea for some, as in your shoe example.  For me, this highlights the lack of a good approach for mixed attributes. If I'm going to use a less than perfect approach, I'd lean towards the least invasive approach, meaning not using mixed attributes at all. This may mean displaying more products with fewer options to keep the option values to a reasonable number.

I guess I just don't see a workable solution to mixed attributes that doesn't pigeon hole the usability of SBA to only a select number of shops.  I open to hearing suggestions, if there is a way that will result in this being usable for most anyone with mixed-attrbiutes.

**Comment by mc12345678** [comment]
As is, DD is lacking for many... That's where change comes into play... With a working copy of SBA in place, DD can be worked on to provide the additional functionality that has been missing from SBA all these years.  If it comes down to incorporating DD into SBA, then so be it, but it may be equivalent to plugins such as Super Orders and Edit Orders... They both work together quite well, but each is maintained separately because of what it takes to address the applicable section(s).  Sure DD hasn't really been updated for several years (a few minor corrections to adjust for ZC version compatibility); however, that doesn't mean that it can't use improvement.

To stop SBA development at single attribute tracking would be to stop short of the possibility of its use. Ideally development is made with the most "complex" condition in mind and then allow all of the other "simpler" to fall into place.  Right now there is a bit of a crossroads, either SBA is edited to work without consideration of DD, DD is edited to work with a broader offering in SBA, or there is a merge of the two.  In the basic code of DD, I don't see anything wrong with it.  It was designed to be malleable to the store's selection.  It currently doesn't offer say multiple methods of mixed attribute selections depending on say the product selected (product A use sequenced dropdown, while product B uses a free for all), but even that is something that can be improved upon moving forward, simply an additional field in some database table that is applicable to the product being tracked (most likely an SBA generated table) with admin control/selection for each product.

Some and total, I say DD is a tool that can be used to bridge the gap that currently exists, but it too needs work to provide the necessary functionality. (Not to pigeon hole the functionality) Like all things with development it will be a "process" to get to the final goal, but ideally as said can provide usability out the gate and add the additional functionality that has been incorporated already with opportunity to grow further... :)

**Comment by jeking928** [comment]
Plugins like Super Orders and Edit Orders do not require each other to work.  It's optional.  They function independently but also together.  SBA, currently, does not function properly without DD. As such, I think that means it should be incorporated in.  While this is certainly more work, it would allow for DD to be cleaned up AND allow SBA to function in the most complex condition, as you describe, which I agree is the best way to approach a plugin as it would allow for use in many more stores.

**Comment by mc12345678** [comment]
FYI, updated the branch to correct the issues identified above.  Still need to properly use the newly created order tracking table when presenting "historical" information, but at least the appropriate information seems to be presented.  Take a look at the mixed attribute orders.php view.  The customid is shown adjacent to the last attribute in the list and is only shown for a variant that exists, not for each individual attribute.
Part of the thought of temporarily keeping DD separate was to improve on DD while still having a functional product... Basically could present this product with the limited features, but with the supported files to provide more functionality with instruction to install the remaining portion of DD.  I mean, in a way don't mind that a program takes off in its own direction, but trying to see how the effort can provide the greatest benefit to the most people rather than just to SBA...

**Comment by jeking928** [comment]
It does look like the customID is displaying correctly in the cart (for the customer) and in the Admin>Order page.  I also see it correctly on account_history_info as well. Inventory updates, I think those issues are all resolved.

DD is probably better in a separate issue?  Getting to the point it's included in steps, does make sense.  Baby steps. And allows it to be improved.

**Comment by mc12345678** [comment]
Basically DD fits into issue #21.  Have already considered a method to support the basic mixed attribute aspect without yet displaying additional SBA table data on each attribute (start adding switches to control operation)

Only thing about the customid at this point is that it is determined/based on live data and when reviewing orders needs to be setup to pull from the new "static" data table that was created rather the live data datatable(s).  There's a few other places into which to intrude (popular purchased items, and anywhere else so associated, but to fully integrate with the base software)

As for the customid itself, I'm thinking that perhaps a switch could/should be added to help with control of the display/output of the customid.  Currently it is used to replace the model information in specific areas, but there may be those that don't want the cutsomid to be displayed, or want both to be displayed.  Perhaps that all is a different issue as well. :)

**Comment by mc12345678** [comment]
The discussion of the breadth of application of the customID I consider complete and applied in issue #23. There are some issues with how it has been implemented which are covered in issue #25. This item is considered closed. Suggest any new discussion of the breadth of application of the customID be initiated in a new issue, and if possible to reference this closed issue.
